genum/gen/enumTemplate.gotmpl

```diff
@@ -87,32 +87,78 @@ func (e {{$enumTypeName}}) String() string {
 	}
 }
 
-// ParseString will return a value as defined in string form.
-func (e {{$enumTypeName}}) ParseString(text string) ({{$enumTypeName}}, error) {
-	switch text {
-	{{- range $val := $values}}
-	case "{{$val.Name}}":
+// Parse{{$enumTypeName}} will attempt to parse the value of a {{$enumTypeName}} from either its string form
+// or any value of a trait flagged with the --parsableByTrait flag
+func Parse{{$enumTypeName}}(input any) ({{$enumTypeName}}, error) {
+	switch input {
+	{{- range $j, $val := $values }}
+	case "{{$val.Name}}"
+	{{- if $.CaseInsensitive -}}
+	, "{{$val.LowerCaseName}}"
+	{{- end -}}
+	{{- range $trait := (index $.Traits $i) -}}
+	{{- if $trait.Parsable -}}
+	{{- $instance := (index $trait.Traits $j) -}}
+	, {{$instance.Value}}
+	{{- end -}}
+	{{- end }}:
 		return {{$val.Name}}, nil
 	{{- end }}
 	default:
-		{{- if $.CaseInsensitive }}
-		switch strings.ToLower(text) {
-		{{- range $val := $values}}
-		case "{{$val.LowerCaseName}}":
-			return {{$val.Name}}, nil
-		{{- end }}
-		}
-		{{- end }}
-		return 0, fmt.Errorf("`%s` is not a valid enum of type {{$enumTypeName}}", text)
+		return 0, fmt.Errorf("`%+v` could not be parsed to enum of type {{$enumTypeName}}", input)
 	}
 }
 
-// ParseStringGeneric calls TypedEnum.ParseString but returns the result
+// ParseGeneric calls TypedEnum.Parse but returns the result
 // in the generic genum.Enum interface. Which is useful when you are only able to work with
 // the un-typed interface.
-func (e {{$enumTypeName}}) ParseStringGeneric(text string) (genum.Enum, error) {
-	return e.ParseString(text)
+func (e {{$enumTypeName}}) ParseGeneric(input any) (genum.Enum, error) {
+	return Parse{{$enumTypeName}}(input)
+}
+
+// Parse{{$enumTypeName}}Int attempts to parse an {{$enumTypeName}} from an int value.
+// If any parsable traits can be converted to an int then they will be parsed if the
+// initial int parsing fails.
+func Parse{{$enumTypeName}}Int(i int) ({{$enumTypeName}}, error) {
```

Comment:
So, I think you got part of what I was saying but not all of it. -- probably all on me. no worries. 

- Let's move the logic to parse individual traits into their respective marshaling functions; it's generated code anyway. But I'm really not a fan of polluting with a bunch of extra public functions.
- as I mentioned before, you do not need to support the use case of parsing an integer anymore. If /when I want that I'll make it a little more intentional. 
- last, you're probably not going to like this; its kinda gross, but tl;dr, types are hard to get right, so I'll give you an example template for the JSON parser. 


```go
	// We always support strings.
	var s string
	if err := json.Unmarshal(data, &s); err == nil {
		var err error
		*e, err = {{$enumTypeName}}(0).Parse(s)
		if err == nil {
			return nil
		}
		{{- range $trait := $.Traits.ParsableStringUnderlying }}
		*e, err = {{$enumTypeName}}(0).Parse({{$trait.TypeRef}}(s))
		if err == nil {
			return nil
		}
		{{- end }}
	}

	// uints:
	{{- if $.Traits.HasParsableUintUnderlying }}
	var uinter64 uint64
	if err := json.Unmarshal(data, &s); err == nil {
		{{- range $trait := $.Traits.ParsableUintUnderlying }}
		*e, err = {{$enumTypeName}}(0).Parse({{$trait.TypeRef}}(uinter64))
		if err == nil {
			return nil
		}
		{{- end }}
	}
	{{- end }}

	// ints:
	{{- if $.Traits.HasParsableIntUnderlying }}
	var sint64 int64
	if err := json.Unmarshal(data, &s); err == nil {
		{{- range $trait := $.Traits.ParsableIntUnderlying }}
		*e, err = {{$enumTypeName}}(0).Parse({{$trait.TypeRef}}(sint64))
		if err == nil {
			return nil
		}
		{{- end }}
	}
	{{- end }}

	// floats64: ( i hate floats)..
	{{- if $.Traits.HasParsableFloat64Underlying }}
	var floater64 float64
	if err := json.Unmarshal(data, &s); err == nil {
		{{- range $trait := $.Traits.ParsableFloat64Underlying }}
		*e, err = {{$enumTypeName}}(0).Parse({{$trait.TypeRef}}(floater64))
		if err == nil {
			return nil
		}
		{{- end }}
	}
	{{- end }}

	// float32:
	{{- if $.Traits.HasParsableFloat32Underlying }}
	var floater32 float32
	if err := json.Unmarshal(data, &s); err == nil {
		{{- range $trait := $.Traits.ParsableFloat32Underlying }}
		*e, err = {{$enumTypeName}}(0).Parse({{$trait.TypeRef}}(floater32))
		if err == nil {
			return nil
		}
		{{- end }}
	}
	{{- end }}

	// native parsing
	{{- range $i, $trait := $.Traits.ParsableImplementsJSONMarshaller }}
	var v{{$i}} {{$trait.TypeRef}}
	if err := json.Unmarshal(data, &s); err == nil {
		*e, err = {{$enumTypeName}}(0).Parse(v{{$i}})
		if err == nil {
			return nil
		}
	}
	{{- end }}

```

So the above is more what I was referring to early on, however, with that example it may be possible, or even preferable to skip the intermediate parsing of int64, uint64, floats, etc to the last version that tries to parse an object directly. It just needs to be tested to ensure it actually works. 

You might be asking why it's so complicated; e.g. why separate units and ints -> if an int is negative uint can't parse it, if a uint is too large int can't parse it. similarly float64 != float32 ; in some cases they cannot be directly cast into one another. I also never accounted for bools.. but kinda hope no one is gonna use a bool for a parsable trait.. that's just an enum of 2 without room for growth. 

Again, all of this complication is why enforcing a string-only is quite nice. But I do like the idea of going a step further. It's just a lot more work.